### ISSUE TITLE
Unable to route request PUT (OOM exception)

### ISSUE THREAD

**Issue by sethrosetter** [couchbase/couchbase-lite-java-core#1298]
Using (Android) 0.0.0-756 + Sync Gateway (1.2.1-4)

This test is failing - https://github.com/couchbaselabs/sync-gateway-tests/blob/master/tests/cbl-large-data.js

From logcat:

```
D/dalvikvm(10960): GC_FOR_ALLOC freed <1K, 32% free 21571K/31431K, paused 35ms, total 35ms
D/dalvikvm(10960): GC_FOR_ALLOC freed <1K, 31% free 21827K/31431K, paused 38ms, total 38ms
I/dalvikvm-heap(10960): Forcing collection of SoftReferences for 4000232-byte allocation
D/dalvikvm(10960): GC_BEFORE_OOM freed 801K, 34% free 21026K/31431K, paused 58ms, total 58ms
E/dalvikvm-heap(10960): Out of memory on a 4000232-byte allocation.
I/dalvikvm(10960): "Acme.Utils.ThreadPool(1)-PooledThread: Acme.Serve.Serve$ServeConnection@41201a40" daemon prio=5 tid=15 RUNNABLE
I/dalvikvm(10960):   | group="TJWS threads" sCount=0 dsCount=0 obj=0x412659d8 self=0x2a41f3c0
I/dalvikvm(10960):   | sysTid=10978 nice=0 sched=0/0 cgrp=apps handle=706530504
I/dalvikvm(10960):   | schedstat=( 8379687008 2742787101 3185 ) utm=516 stm=321 core=0
I/dalvikvm(10960):   at com.fasterxml.jackson.core.util.ByteArrayBuilder.toByteArray(ByteArrayBuilder.java:~118)
I/dalvikvm(10960):   at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsBytes(ObjectMapper.java:2807)
I/dalvikvm(10960):   at com.couchbase.lite.support.RevisionUtils.asCanonicalJSON(RevisionUtils.java:159)
I/dalvikvm(10960):   at com.couchbase.lite.store.SQLiteStore.add(SQLiteStore.java:1398)
I/dalvikvm(10960):   at com.couchbase.lite.Database.put(Database.java:1837)
I/dalvikvm(10960):   at com.couchbase.lite.Database.putRevision(Database.java:1787)
I/dalvikvm(10960):   at com.couchbase.lite.Database.putRevision(Database.java:1778)
I/dalvikvm(10960):   at com.couchbase.lite.router.Router.update(Router.java:2138)
I/dalvikvm(10960):   at com.couchbase.lite.router.Router.update(Router.java:2191)
I/dalvikvm(10960):   at com.couchbase.lite.router.Router.do_PUT_Document(Router.java:2227)
I/dalvikvm(10960):   at java.lang.reflect.Method.invokeNative(Native Method)
I/dalvikvm(10960):   at java.lang.reflect.Method.invoke(Method.java:511)
I/dalvikvm(10960):   at com.couchbase.lite.router.Router.start(Router.java:568)
I/dalvikvm(10960):   at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
I/dalvikvm(10960):   at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
I/dalvikvm(10960):   at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2251)
I/dalvikvm(10960):   at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2176)
I/dalvikvm(10960):   at Acme.Serve.Serve$ServeConnection.run(Serve.java:1988)
I/dalvikvm(10960):   at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1223)
I/dalvikvm(10960):   at java.lang.Thread.run(Thread.java:856)
I/dalvikvm(10960):
W/Router  (10960): Router unable to route request to do_PUT_Document
W/Router  (10960): java.lang.reflect.InvocationTargetException
W/Router  (10960): 	at java.lang.reflect.Method.invokeNative(Native Method)
W/Router  (10960): 	at java.lang.reflect.Method.invoke(Method.java:511)
W/Router  (10960): 	at com.couchbase.lite.router.Router.start(Router.java:568)
W/Router  (10960): 	at com.couchbase.lite.listener.LiteServlet.service(LiteServlet.java:141)
W/Router  (10960): 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:802)
W/Router  (10960): 	at Acme.Serve.Serve$ServeConnection.runServlet(Serve.java:2251)
W/Router  (10960): 	at Acme.Serve.Serve$ServeConnection.parseRequest(Serve.java:2176)
W/Router  (10960): 	at Acme.Serve.Serve$ServeConnection.run(Serve.java:1988)
W/Router  (10960): 	at Acme.Utils$ThreadPool$PooledThread.run(Utils.java:1223)
W/Router  (10960): 	at java.lang.Thread.run(Thread.java:856)
W/Router  (10960): Caused by: java.lang.OutOfMemoryError
W/Router  (10960): 	at com.fasterxml.jackson.core.util.ByteArrayBuilder.toByteArray(ByteArrayBuilder.java:118)
W/Router  (10960): 	at com.fasterxml.jackson.databind.ObjectMapper.writeValueAsBytes(ObjectMapper.java:2807)
W/Router  (10960): 	at com.couchbase.lite.support.RevisionUtils.asCanonicalJSON(RevisionUtils.java:159)
W/Router  (10960): 	at com.couchbase.lite.store.SQLiteStore.add(SQLiteStore.java:1398)
W/Router  (10960): 	at com.couchbase.lite.Database.put(Database.java:1837)
W/Router  (10960): 	at com.couchbase.lite.Database.putRevision(Database.java:1787)
W/Router  (10960): 	at com.couchbase.lite.Database.putRevision(Database.java:1778)
W/Router  (10960): 	at com.couchbase.lite.router.Router.update(Router.java:2138)
W/Router  (10960): 	at com.couchbase.lite.router.Router.update(Router.java:2191)
W/Router  (10960): 	at com.couchbase.lite.router.Router.do_PUT_Document(Router.java:2227)
W/Router  (10960): 	... 10 more
D/dalvikvm(10960): GC_FOR_ALLOC freed 12204K, 72% free 9073K/31431K, paused 35ms, total 37ms
I/dalvikvm-heap(10960): Grow heap (frag case) to 12.724MB for 4000232-byte allocation
D/dalvikvm(10960): GC_CONCURRENT freed 5K, 59% free 12974K/31431K, paused 17ms+6ms, total 48ms
D/dalvikvm(10960): WAIT_FOR_CONCURRENT_GC blocked 19ms
D/dalvikvm(10960): GC_CONCURRENT freed 1K, 54% free 14627K/31431K, paused 17ms+17ms, total 83ms
```

**Comment by sethrosetter** [comment]
Using API 16 ARM emulator

**Comment by sethrosetter** [comment]
To Repro (Using arm API 16):
- Create a Android db
- Add 10 4mb docs
- Crash with OOM

**Comment by sethrosetter** [comment]
I need to verify against newer arm emulator.

**Comment by hideki** [comment]
I added following test which inserts 4MB doc 10 times through Native API. The test passed with ARM API 16 emulator. I will set this test's milestone to `1.4`.
https://github.com/couchbase/couchbase-lite-android/commit/5c014b16cb1d4099e1379a7014cd8c85722ab933

Note:
1. I marked this test as manual test because this inserting large data.  This kind of test usually makes other test unstable.
2. This test does not do retrieve all documents which are just inserted. Retrieving all 4MB docs also cause OOM, but I am not sure if this test is feasible test with limited memory device.
